hbase-server/src/main/java/org/apache/hadoop/hbase/master/HbckChecker.java

```diff
@@ -0,0 +1,220 @@
+    } catch (IOException e) {
+      LOG.warn("Faile to load the regions from filesystem", e);
+    }
+    saveCheckResultToSnapshot();
+    running = false;
+  }
+
+  private void saveCheckResultToSnapshot() {
+    synchronized (orphanRegionsOnRSSnapshot) {
+      orphanRegionsOnRSSnapshot.clear();
+      orphanRegionsOnRS.entrySet()
+          .forEach(e -> orphanRegionsOnRSSnapshot.put(e.getKey(), e.getValue()));
+    }
+    synchronized (orphanRegionsOnFSSnapshot) {
+      orphanRegionsOnFSSnapshot.clear();
+      orphanRegionsOnFSSnapshot.addAll(orphanRegionsOnFS);
+    }
+    synchronized (inconsistentRegionsSnapshot) {
+      inconsistentRegionsSnapshot.clear();
+      inconsistentRegions.entrySet()
+          .forEach(e -> inconsistentRegionsSnapshot.put(e.getKey(), e.getValue()));
+    }
+  }
+
+  private void loadRegionsFromInMemoryState() {
+    List<RegionState> regionStates =
+        master.getAssignmentManager().getRegionStates().getRegionStates();
+    for (RegionState regionState : regionStates) {
+      RegionInfo regionInfo = regionState.getRegion();
+      HbckRegionInfo.MetaEntry metaEntry =
+          new HbckRegionInfo.MetaEntry(regionInfo, regionState.getServerName(),
+              regionState.getStamp());
+      regionInfoMap.put(regionInfo.getEncodedName(), new HbckRegionInfo(metaEntry));
+    }
+  }
+
+  private void loadRegionsFromRSReport() {
+    Map<ServerName, Set<byte[]>> rsReports = master.getAssignmentManager().getRSReports();
+    for (Map.Entry<ServerName, Set<byte[]>> entry : rsReports.entrySet()) {
+      ServerName serverName = entry.getKey();
+      for (byte[] regionName : entry.getValue()) {
+        String encodedRegionName = RegionInfo.encodeRegionName(regionName);
+        HbckRegionInfo hri = regionInfoMap.get(encodedRegionName);
+        if (hri == null) {
+          orphanRegionsOnRS.put(encodedRegionName, serverName);
+          continue;
+        }
+        hri.addServer(hri.getMetaEntry(), serverName);
+      }
+    }
+
+    for (Map.Entry<String, HbckRegionInfo> entry : regionInfoMap.entrySet()) {
+      String encodedRegionName = entry.getKey();
+      HbckRegionInfo hri = entry.getValue();
+      ServerName locationInMeta = hri.getMetaEntry().getRegionServer();
+      if (hri.getDeployedOn().size() == 0) {
+        // Master thought this region opened, but no regionserver reported it.
+        inconsistentRegions.put(encodedRegionName, new Pair<>(locationInMeta, new LinkedList<>()));
+      } else if (hri.getDeployedOn().size() > 1) {
+        // More than one regionserver reported opened this region
+        inconsistentRegions.put(encodedRegionName, new Pair<>(locationInMeta, hri.getDeployedOn()));
+      } else if (!hri.getDeployedOn().get(0).equals(locationInMeta)) {
+        // Master thought this region opened on Server1, but regionserver reported Server2
+        inconsistentRegions.put(encodedRegionName, new Pair<>(locationInMeta, hri.getDeployedOn()));
+      }
+    }
+  }
```

Comment:
Is this info going to be stale by the time we go to display it? Will it be reporting problems or transient conditions?